### PR TITLE
fix: resolve infinite loop in _white_list_collection

### LINKED ISSUE #230: Admin won't be able to remove certain collections from the whitelist on L2
From https://codehawks.cyfrin.io/c/2024-07-ark-project/s/124

Admin won't be able to remove collections from the whitelist on L2 in certain cases. The issue stems from the fact that there's a missing line in the `_white_list_collection` function in the `bridge.cairo` contract that will make the function loop indefinitely if the collection to be removed is after 2nd place in the linked list.

Removing a collection by an admin should always be possible. This will have an even bigger impact if a malicious collection finds its way into the whitelist either by accident or at a time that the whitelist has been off.

### PR DESCRIPTION
**Pull request by JoE11-y**
Added missing line issue causing the issue where the admin was unable to remove certain collections. Added a unit test to verify correct behavior and prevent future regressions.

Resolves #230

### CODE REVIEW

File: apps/blockchain/starknet/src/tests/bridge_t.cairo [hunk 1]
```diff
@@ -831,6 +831,102 @@ mod tests {
 
     }
 
+    #[test]
```
Comment (ptisserand) [inline apps/blockchain/starknet/src/tests/bridge_t.cairo:834]: Adding event checking is a good idea, but you should update `whitelist_collection_is_updated_when_collection_is_removed` instead.

**Review by ptisserand — CHANGES_REQUESTED** [review]

File: apps/blockchain/starknet/src/tests/bridge_t.cairo (hunk 1 above)
Comment (JoE11-y) [reply apps/blockchain/starknet/src/tests/bridge_t.cairo]: Just saw it and I've updated.
Thanks @ptisserand

File: apps/blockchain/starknet/src/tests/bridge_t.cairo [hunk 2]
```diff
@@ -831,6 +844,53 @@ mod tests {
...
+    //             bridge::Event::CollectionWhiteListUpdated(
+    //                 bridge::CollectionWhiteListUpdated {
+    //                     collection: collection3,
+    //                     enabled: false,
+    //                 }
+    //             )
+    //         ),
+    //     ]);
+    // }
+
```
Comment (ptisserand) [inline apps/blockchain/starknet/src/tests/bridge_t.cairo:893]: Please remove dead code

File: apps/blockchain/starknet/src/tests/bridge_t.cairo [hunk 3]
```diff
@@ -694,81 +694,94 @@ mod tests {
         let collection1 = starknet::contract_address_const::<'collection1'>();
         let collection2 = starknet::contract_address_const::<'collection2'>();
         let collection3 = starknet::contract_address_const::<'collection3'>();
-
+        let collection4 = starknet::contract_address_const::<'collection4'>();
+        let collection5 = starknet::contract_address_const::<'collection5'>();
+  
+        let mut spy = spy_events(SpyOn::One(bridge_address));
```
Comment (ptisserand) [inline apps/blockchain/starknet/src/tests/bridge_t.cairo:700]: Could you please add also event checking for at least 1 call

File: apps/blockchain/starknet/src/tests/bridge_t.cairo [hunk 4]
```diff
@@ -694,81 +694,94 @@ mod tests {
...
+        assert!(bridge.is_white_listed(collection5), "Collection1 should be whitelisted");
 
         start_prank(CheatTarget::One(bridge_address), BRIDGE_ADMIN);
         bridge.white_list_collection(collection1, false);
-        bridge.white_list_collection(collection3, false);
+        bridge.white_list_collection(collection4, false);
         stop_prank(CheatTarget::One(bridge_address));
-        let white_listed = bridge.get_white_listed_collections();
-        assert!(white_listed.is_empty(), "White list shall be empty");
+        assert_eq!(white_listed.len(), 2, "White list shall contain 2 elements");
```
Comment (ptisserand) [inline apps/blockchain/starknet/src/tests/bridge_t.cairo:741]: Test is failing, you forgot this line here:
```cairo
        let white_listed = bridge.get_white_listed_collections();
```

**Review by ptisserand — CHANGES_REQUESTED** [review]
Thanks, some minor changes and it should be good to be merged

File: apps/blockchain/starknet/src/tests/bridge_t.cairo (hunk 3 above)
Comment (JoE11-y) [reply apps/blockchain/starknet/src/tests/bridge_t.cairo:700]: Yes I've added this.

File: apps/blockchain/starknet/src/tests/bridge_t.cairo [hunk 5]
```diff
@@ -694,81 +694,149 @@ mod tests {
...
         assert_eq!(*white_listed.at(0), collection1, "Wrong collection address in white list");
         assert_eq!(*white_listed.at(1), collection2, "Wrong collection address in white list");
         assert_eq!(*white_listed.at(2), collection3, "Wrong collection address in white list");
+        assert_eq!(*white_listed.at(3), collection4, "Wrong collection address in white list");
+        assert_eq!(*white_listed.at(4), collection5, "Wrong collection address in white list");
         assert!(bridge.is_white_listed(collection1), "Collection1 should be whitelisted");
         assert!(bridge.is_white_listed(collection2), "Collection1 should be whitelisted");
         assert!(bridge.is_white_listed(collection3), "Collection1 should be whitelisted");
+        assert!(bridge.is_white_listed(collection4), "Collection1 should be whitelisted");
+        assert!(bridge.is_white_listed(collection5), "Collection1 should be whitelisted");
```
Comment (ptisserand) [inline apps/blockchain/starknet/src/tests/bridge_t.cairo:721]: Could you please move this line 
```cairo
let mut spy = spy_events(SpyOn::One(bridge_address));
```
just before
```cairo
        start_prank(CheatTarget::One(bridge_address), BRIDGE_ADMIN);
        bridge.white_list_collection(collection3, false);
        stop_prank(CheatTarget::One(bridge_address));
```
to ensure that `assert_emitted` only check events from:
```cairo
bridge.white_list_collection(collection3, false);
```

**Review by ptisserand — CHANGES_REQUESTED** [review]

File: apps/blockchain/starknet/src/tests/bridge_t.cairo (hunk 5 above)
Comment (JoE11-y) [reply apps/blockchain/starknet/src/tests/bridge_t.cairo:721]: alright @ptisserand

**Review by ptisserand — APPROVED** [review]
Thanks a lot for your contribution!
LGTM

**Pull request merged**